pkg/asset/manifests/gcp/cluster.go

```diff
@@ -137,32 +137,35 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
 // only use zones if both the compute and control plane node availability zones exist.
 func findFailureDomains(installConfig *installconfig.InstallConfig) []string {
 	zones := sets.New[string]()
-	defaultMachinePlatformDefined := false
-	if gcpPlatform := installConfig.Config.Platform.GCP; gcpPlatform != nil && gcpPlatform.DefaultMachinePlatform != nil {
-		defaultMachinePlatformDefined = true
-		for _, zone := range gcpPlatform.DefaultMachinePlatform.Zones {
+
+	var controlPlaneZones, computeZones []string
+	if installConfig.Config.ControlPlane.Platform.GCP != nil {
+		controlPlaneZones = installConfig.Config.ControlPlane.Platform.GCP.Zones
+	}
+
+	if installConfig.Config.Compute[0].Platform.GCP != nil {
+		computeZones = installConfig.Config.Compute[0].Platform.GCP.Zones
+	}
+
+	def := installConfig.Config.GCP.DefaultMachinePlatform
+	if def != nil && len(def.Zones) > 0 {
+		for _, zone := range def.Zones {
+			zones.Insert(zone)
+		}
+
+		for _, zone := range controlPlaneZones {
 			zones.Insert(zone)
 		}
 
-		if installConfig.Config.ControlPlane.Platform.GCP != nil {
-			for _, zone := range installConfig.Config.ControlPlane.Platform.GCP.Zones {
-				zones.Insert(zone)
-			}
+		for _, zone := range computeZones {
+			zones.Insert(zone)
 		}
-		if installConfig.Config.Compute[0].Platform.GCP != nil {
-			for _, zone := range installConfig.Config.Compute[0].Platform.GCP.Zones {
-				zones.Insert(zone)
-			}
+	} else if len(controlPlaneZones) > 0 && len(computeZones) > 0 {
```

Comment:
@patrickdillon I think this part was always a bit confusing to me. Why do both compute and control plane node zones have to be non-empty but the values in them don't need to match? If the zones for one of them we would then default to using all availability zones (or failure domains as the recipient field suggests).

Reply:
If the machinepool doesn't specify the AZ that means use any/all AZs. So if either compute or control plane don't set AZs, we would need all AZs set for the cluster network. The only case where we want to limit (not use all) AZs is when the AZs are explicitly set for both pools.

Reply:
And this is ok that we don't check that those AZs match in this case (no overlap between them)?

Reply:
Technically it is fine If users specify compute and control plane nodes are in disjoint sets of AZs. 

Really most of this is an academic consideration because there are so few zones (typically three) that there will most likely be overlap.

Reply:
Thanks for clarifying !